uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/rules/FactoryResetChromeRule.java

```diff
@@ -0,0 +1,103 @@
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.client.ui.automation.rules;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.text.TextUtils;
+
+import androidx.test.core.app.ApplicationProvider;
+
+import com.microsoft.identity.client.ui.automation.browser.BrowserChrome;
+import com.microsoft.identity.client.ui.automation.logging.Logger;
+import com.microsoft.identity.client.ui.automation.utils.AdbShellUtils;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * A Test Rule to downgrade Chrome and WebView to factory default version. The newer versions of
+ * Chrome & WebView seem to be not so suitable for automation and we run into resource id not found
+ * error. Basically it seems that Chrome rolled out an update due to which automation frameworks are
+ * not able to find elements on the web-page (at least this is the case for ESTS UX page). To
+ * mitigate this, if we find that Chrome or WebView version on the device is above the known version
+ * suitable for automation, then we would factory reset Chrome.
+ */
+public class FactoryResetChromeRule implements TestRule {
+
+    public static final String TAG = FactoryResetChromeRule.class.getSimpleName();
+
+    private final static int CHROME_MAJOR_VERSION_SUITABLE_FOR_AUTOMATION = 74;
+    private final static int WEB_VIEW_MAJOR_VERSION_SUITABLE_FOR_AUTOMATION = 74;
+    private final static String WEB_VIEW_PACKAGE_NAME = "com.google.android.webview";
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                final String chromeVersion = getPackageMajorVersion(BrowserChrome.CHROME_PACKAGE_NAME);
+                if (!TextUtils.isEmpty(chromeVersion) &&
+                        Integer.parseInt(chromeVersion) > CHROME_MAJOR_VERSION_SUITABLE_FOR_AUTOMATION) {
+                    Logger.w(TAG, "Chrome version on the device is higher than the known version suitable for automation. " +
+                            "We are going to attempt to factory reset Chrome and hope that it will give us the desired version.");
+                    // adb uninstall does factory reset for default apps
+                    AdbShellUtils.removePackage(BrowserChrome.CHROME_PACKAGE_NAME);
+                }
```

Comment:
Should we check this and see if it worked?

Reply:
We can log the new version after downgrade (if it still not the version we want then I am planning to just be optimistic and proceed with the test anyway -- because may be it starts working one day)

Reply:
Hm.  If we're insisting on a version in order to run the tests, and then we don't get it but it succeeds anyway, can we get the framework to give us a signal that our assumptions have changed *other* than just failing the test?

Reply:
We would have the logs but won't have an alert. We can look into to see how to send the alert. Another thing we can do is that instead of failing or passing the test, we can abort the test if the version we need is not there using assumptions: https://junit.org/junit5/docs/5.0.0/api/org/junit/jupiter/api/Assumptions.html

Reply:
How likely is it that any given engineer is going to pay attention to any of this?

Reply:
Not likely...abort is just ignoring the test and likely the engineer won't pay much attention. I don't currently have a mechanism to pass the test and still throw an alert (it might be possible but would require investigation). For the time being then let's just fail the test if we don't get the version that we are looking for even after downgrade. Test failure is something that release engineer should be looking at.

Reply:
Addressed here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1204/commits/1561de7500afe0b01c7f865d04b37bdf22a4a4b5